### ISSUE TITLE
Deprecated re-exports are ignored

### ISSUE THREAD

**Issue by sgrif** [rust-lang/rust#47236]
The following code compiles with no warnings, even though it should emit a deprecation warning. Both Diesel and Futures have mistakenly used this to deprecate items moved to a new location, without realizing that users would never see a deprecation warning.

```rust
pub mod foo {
    #[deprecated(since = "0.1.0")]
    pub use std::option::Option;
}

mod bar {
    use foo::Option;
}
```

**Comment by petrochenkov** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/23937 (except it's about the user-facing `deprecated` attribute and https://github.com/rust-lang/rust/issues/23937 is about internal stability attributes in the standard library, but the mechanism is the same).

**Comment by petrochenkov** [comment]
Also similar to https://github.com/rust-lang/rust/issues/15728, both can be fixed by moving stability/deprecation checking to name resolution stage.

**Comment by dtolnay** [comment]
Closing as a duplicate of #30827, which has some discussion from compiler team members on how to solve this.
